### ISSUE TITLE
Thoughts about SSR

### ISSUE THREAD

**Issue by Paratron** [pmndrs/valtio#14]
This is not an issue but just some thoughts I wanted to share.

I saw the design of valtio a few days ago on twitter and liked it a lot. Sure, you still need to work carefully with anything that updates the state or you will very quickly land in where-did-my-state-update-from-hell 😅

Anyways - one thing that stopped my mind for a second was: how would this work in SSR?
The problem is that one cannot just have a global state object hanging somewhere in a JS file to be imported and manipulated, because in SSR its a bit like you have the app running several times in parallel but each one needs a separate state.

I had the idea that this can be accomplished by wrapping valtios `proxy` and `useProxy` just a bit and routing the state object through a context.

Here is a CodeSandbox that runs both in SSR as well as in the browser:

https://codesandbox.io/s/mystifying-heyrovsky-eu0ek?file=/src/App.js

In SSR, the state remains a plain object. Its important that all data necessary to render the app are fetched before starting to render.

When the app notices its running in a browser context, the state will be "activated" and becomes a proxy object that can be manipulated at will and causes UI updates. :)

This is just a rough idea written down in a couple of minutes - but in theory it should also work with a nextJS app.

**Comment by dai-shi** [comment]
Yeah, global state is not SSR friendly.
We have a long discussion in zustand. https://github.com/pmndrs/zustand/issues/182

Just had a quick look at the code, looks pretty feasible. Seems much easier than zustand as state is self contained.
There might be some edge cases to consider, but overall it seems good and promising.

It's really nice to have people come up with various patterns base on the primitives provided by the library. 👏

**Comment by AjaxSolutions** [comment]
@dai-shi Do you see any potential issues with using Valtio in Next.js?

Next.js supports IE11 whereas Valtio (I believe) does not support it.

Otherwise, there should be no issues.  Is my assumption correct?

**Comment by Paratron** [comment]
When you add a react context to keep the state, it should work. Valtio as it is cannot work with nextJS since the state of individually requested pages would get mixed up.

**Comment by AjaxSolutions** [comment]
Thanks, but IMO this needs to be better explained and documented including an example showing how to use Valtio with Next.js.

**Comment by Paratron** [comment]
Well, you are aware that this is more like a proof of concept and not a widely used library, right?

Feel free to add docs and examples 👍

**Comment by AjaxSolutions** [comment]
> When you add a react context to keep the state, it should work.

This is how [MobX](https://github.com/vercel/next.js/blob/canary/examples/with-mobx/pages/_app.js) stores are used with Next.js.  MobX stores are stable between renders.

**Comment by dai-shi** [comment]
@AjaxSolutions 
It should basically work if you take care of it.
We haven't reached to some best practices yet. So, your work on this would be very valuable.
The way @Paratron did is a good option. The nextjs with mobx example would be good too. You might want to try using useLocalProxy in utils.

**Comment by dai-shi** [comment]
> You might want to try using useLocalProxy in utils.

Note we dropped useLocalProxy.

Let me close this issue for now. Suggestions are welcome.
If someone is interested in creating nextjs example, that's good.
